### ISSUE TITLE
Implement `.src`

### ISSUE THREAD

**Issue by phated** [gulpjs/vinyl-prepare#4]
As per https://github.com/gulpjs/vinyl-prepare/issues/2#issuecomment-242515652
- fold `filter-since` (probably move filtering into `glob-stream`) and `wrap-with-vinyl-file` into a single stream
- transplant that to `vinyl-prepare`
- apply options per file

**Comment by erikkemperman** [comment]
@phated I have a rough draft:

https://github.com/erikkemperman/vinyl-fs/commits/refactor-prepare-read
https://github.com/erikkemperman/vinyl-prepare/commits/refactor-prepare-read

It needs tests, in particular for the per-file evaluation of options for vinyl-fs and in general for vinyl-prepare, but I'd prefer to not do those until main changes are more or less settled (or not).

Some things I should mention:

The first commit to vinyl-fs is just something cosmetic, and subjective -- actually I kind of started the rest on top of it by accident. I left it in here for now but could easily get rid of it or submit it separately.

I wasn't completely sure if the glob-stream call ought to be in vinyl-prepare or in vinyl-fs. I picked the former, mostly to keep the signature of `prepare.read` the same as `vfs.src`, in the same way as was done with `prepare.write` and `vfs.dest`.

I left the processing of options like `read`, `buffer` and `stripBOM` out of prepare  -- and in the readXxx modules -- for now. On the one hand it seems to me that all options processing should be centralized, in vinyl-prepare and available for separate use, but I agree that my attempt to achieve that, with the changes to `prepare.write` I linked earlier, is not great... I'll see if I can come up with something better.

There are some options still "global", not evaluated per file:
- the options passed on to through2;
- the options passed on to glob-stream (*);
- `passthrough` for obvious reasons;
- `sourcemaps` for the same reason as in `dest`, pending the issue in gulp-sourcemaps;
- `followSymlinks` (**)

(*) Function options to glob-stream could be evaluated per file,  as discussed in https://github.com/gulpjs/glob-stream/issues/66 , but they would have a different signature from "internal" vinyl-fs options. I think it would make sense, but documentation would have make sure to avoid confusion there -- e.g. that `since` would actually be a glob-stream option with its own contract.

(**) I found that last one a bit tricky. I suppose it could be per-file, but I'm not sure how people use this option and what they would expect. So I propose leaving it like this for now, pending https://github.com/gulpjs/vinyl-fs/issues/205 at least.

Finally, there are some potential shortcuts for constant-value options (as opposed to functions). For instance, the whole `readContents` stage can be excised from the pipeline if `options.read === false`, instead of finding over and over for each of the files that we're going to have to skip it. I'm not sure if this kind of thing is a worth some minor complications in the code, though.

**Comment by phated** [comment]
@erikkemperman have you had a chance to dig into this any more?  Not sure how far down the rabbit-hole it sent you.

**Comment by erikkemperman** [comment]
Not too far. Well, no further, anyway. I've shelved this whilst waiting on the PR to node-glob. But that might not go anywhere, or not anytime soon, so if it helps things along I could go ahead and move the glob-stream call and symlink resolution back up in vinyl-fs for the time being?

**Comment by phated** [comment]
@erikkemperman I think those should be moved back into vinyl-fs.

I'm envisioning the pipeline to look something like:
```js
globStream()
  .pipe(vinylPrepare.src())
  .pipe(readContents())
```

**Comment by erikkemperman** [comment]
@phated I've updated my branches to move the glob-stream call and symlink resolution back into vinyl-fs. 

https://github.com/erikkemperman/vinyl-fs/commits/refactor-prepare-read
https://github.com/erikkemperman/vinyl-prepare/commits/refactor-prepare-read

The pipeline you envisioned is not exactly what I ended up with; it needs an extra stage for wrapping in a vinyl object and resolving the symlinks, in between glob-stream and vinyl-prepare.

Basically all that's left for `prepare.read` to do now is filter on timestamps. And come to think of it, relative to the "core business" of `src` it is actually more post-processing than preparation :-)

I might keep my previous versions of this, with glob-stream and symlink stuff in vinyl-prepare, around for a while -- if/when my node-glob PR goes anywhere I might propose moving them once again.

**Comment by phated** [comment]
@erikkemperman were you not happy with constructing the Vinyl object inside vinyl-prepare?  I figured that was the logical place for it to go.

**Comment by erikkemperman** [comment]
Well, yeah, I agree. But that's done after `fs.realpath` calls, if any, and I thought your point was to keep fs stuff out of prepare?

**Comment by phated** [comment]
Just jam the `realpath` information onto the object `glob-stream` emits.  Pass that object to `vinylPrepare.src` and construct the Vinyl object.

**Comment by erikkemperman** [comment]
I'll take a stab at it, maybe today. Wanted to take a closer look at the realpath stuff anyway -- I haven't followed the breadcrumbs there in detail yet, but was curious about the need for recursion.

**Comment by erikkemperman** [comment]
BTW As I mentioned earlier, these complications would not be necessary if we could pass a custom fs adapter (object with `readdir`, `stat`, ...) to node-glob. That same option could then work in vinyl-prepare, allowing peers of vinyl-fs to pass a compact fs abstraction and keep globbing logic and symlink resolution central/shared.

But I haven't heard back yet. Perhaps you don't agree this would be a good idea, but if you do I wonder if it would help if there's some more "weight" behind that request on node-glob?

**Comment by phated** [comment]
@erikkemperman that doesn't fully solve the problem either.  Some vinyl adapters don't want to do any of those things.  Coupling the Vinyl object creation to all of that seems like a very leaky abstraction.

**Comment by phated** [comment]
Globbing/stating/reading/etc should not be handled by this library.  It should just turn a stream of object (given options) into a stream of vinyl objects

**Comment by erikkemperman** [comment]
Of course `readdir`, for instance, at that point is just a naming convention for telling glob-stream/node-glob how to translate `*` into path components. Adapters that don't support that wouldn't need to provide `readdir`.

And conversely I don't see how those that do want to support asterisks would not have something (conceptually) like `readdir`.

But I guess you might have adapters in mind that I simply lack the imagination to consider :-)

All right, so I'll try to get the vinyl constructor back into prepare, and leave it at that for this PR.

**Comment by erikkemperman** [comment]
@phated Updated my branches to move the vinyl constructor into vinyl-prepare. I'm not sure if I had parsed your remark about "jamming the realpath information" correctly, though -- this seems like quite a vinyl-fs specific property to be expecting in vinyl-prepare:
https://github.com/erikkemperman/vinyl-prepare/commit/3558db36478500d5508431b751947d7444003d8a#diff-168726dbe96b3ce427e7fedce31bb0bcR26

Also, while this doesn't have the glob-stream call in prepare, it is still sort of particular to glob-stream's output in what vinyl-prepare expects coming in.

As an aside: I had a bit of a red-faced moment earlier... Getting more or less familiar with these codebases I can't help getting ideas, for better or worse, but it occurred to me it might get a bit distracting/annoying/tiresome at your end... I trust you understand it's all meant to be constructive and positive, but please let me know if I should dial it down :-)

**Comment by erikkemperman** [comment]
Forgot to mention: I had already left the `resolveSymlink` option global (not evaluated per file) because I wasn't sure how users would expect that to work. But with the vinyl constructor now detached from the symlink resolution that will be harder to change, if the idea is for option-function callbacks is still to always receive a vinyl as their argument.
